WalletWasabi.Tests/UnitTests/WabiSabi/Client/BobClientTests.cs

```diff
@@ -0,0 +1,91 @@
+using Xunit;
+
+namespace WalletWasabi.Tests.UnitTests.WabiSabi.Client
+{
+	public class BobClientTests
+	{
+		[Fact]
+		public async Task RegisterOutputTestAsync()
+		{
+			var config = new WabiSabiConfig { MaxInputCountByRound = 1 };
+			var round = WabiSabiFactory.CreateRound(config);
+			using Arena arena = await WabiSabiFactory.CreateAndStartArenaAsync(config, round);
+			await arena.TriggerAndWaitRoundAsync(TimeSpan.FromMinutes(1));
+
+			var km = ServiceFactory.CreateKeyManager("");
+			var key = BitcoinFactory.CreateHdPubKey(km);
+			SmartCoin coin1 = BitcoinFactory.CreateSmartCoin(key, Money.Coins(1m));
+			var outpoint = coin1.OutPoint;
+
+			var mockRpc = new Mock<IRPCClient>();
+			mockRpc.Setup(rpc => rpc.GetTxOutAsync(outpoint.Hash, (int)outpoint.N, true))
+				.ReturnsAsync(new NBitcoin.RPC.GetTxOutResponse
+				{
+					IsCoinBase = false,
+					Confirmations = coin1.Height,
+					TxOut = coin1.TxOut,
+				});
+
+			CredentialPool amountCredential = new();
+			CredentialPool weightCredential = new();
+
+			await using var coordinator = new ArenaRequestHandler(config, new Prison(), arena, mockRpc.Object);
+			var aliceArenaClient = new ArenaClient(round.AmountCredentialIssuerParameters, round.WeightCredentialIssuerParameters, amountCredential, weightCredential, coordinator, new InsecureRandom());
+			var bobArenaClient = new ArenaClient(round.AmountCredentialIssuerParameters, round.WeightCredentialIssuerParameters, amountCredential, weightCredential, coordinator, new InsecureRandom());
+			Assert.Equal(Phase.InputRegistration, round.Phase);
+
+			var bitcoinSecret = km.GetSecrets("", coin1.ScriptPubKey).Single().PrivateKey.GetBitcoinSecret(Network.Main);
+			var aliceClient = await AliceClient.CreateNewAsync(aliceArenaClient, new[] { coin1.Coin }, bitcoinSecret, round.Id, round.Hash, round.FeeRate);
+
+			Task confirmationTask = aliceClient.ConfirmConnectionAsync(TimeSpan.FromSeconds(3), CancellationToken.None);
+
+			await arena.TriggerAndWaitRoundAsync(TimeSpan.FromMinutes(1));
+			await confirmationTask;
+
+			Assert.Equal(Phase.ConnectionConfirmation, round.Phase);
+
+			await arena.TriggerAndWaitRoundAsync(TimeSpan.FromMinutes(1));
+
+			Assert.Equal(Phase.OutputRegistration, round.Phase);
+
+			using var destinationKey1 = new Key();
+			using var destinationKey2 = new Key();
+			using var destinationKey3 = new Key();
+			using var destinationKey4 = new Key();
+
+			var bobClient = new BobClient(round.Id, bobArenaClient);
+			await bobClient.RegisterOutputAsync(Money.Coins(0.25m), destinationKey1.PubKey.WitHash.ScriptPubKey);
+
+			await bobClient.RegisterOutputAsync(Money.Coins(0.25m), destinationKey2.PubKey.WitHash.ScriptPubKey);
+
+			await bobClient.RegisterOutputAsync(Money.Coins(0.25m), destinationKey3.PubKey.WitHash.ScriptPubKey);
+
+			await bobClient.RegisterOutputAsync(amountCredential.Valuable.Sum(c => c.Amount.ToMoney()), destinationKey4.PubKey.WitHash.ScriptPubKey);
+
+			Assert.Empty(amountCredential.Valuable);
```

Comment:
I would prefer if relationship between the credential amounts and the registration amounts was not included since it is making assertions more about how the credential pool works than how the client uses it, and this is probably not going to keep working like that for very long

Reply:
ACK